Murmur3fix

Claudenw opened the pull request:
CODEC-264: Fixed sign extensions.

Fix for both 32 and 128 bit hashes.
Additional test cases added including comments re: source of tests.

aherbert commented on src/main/java/org/apache/commons/codec/digest/MurmurHash3.java:
```diff
@@ -393,7 +401,7 @@ public static long hash64(final byte[] data, final int offset, final int length,
 	 */
 	public static long[] hash128(final String data) {
-		final byte[] origin = data.getBytes();
+		final byte[] origin = data.getBytes(StandardCharsets.UTF_8);
```
Why the charset conversion? Changing away from using the default charset is not related to the other changes in this PR.

Claudenw replied:
It showed up in a bug report -- reverting.

aherbert commented on src/main/java/org/apache/commons/codec/digest/MurmurHash3.java:
```diff
@@ -602,15 +613,19 @@ public final void add(final byte[] data, int offset, final int length) {
 			switch (tailLen) {
 			case 3:
-				k1 ^= tail[2] << 16;
+				k1 ^= (tail[2] & UBYTE_MASK) << 16;
```
In the change to `hash32` you switched to using `|=` operators from `^=` but not here.

In both cases it makes no difference as k1 is zero and `^=` will produce the equivalent. I suggest changing the `|=` operator back to `^=`. This will minimise the changes in the PR.

A performance test against an updated algorithm that exclusively uses `|=` would justify a change. Otherwise just leave it as is. I'd expect no difference anyway given this is in the `end()` method and will be a minority of the runtime for the overall hash method.

Claudenw replied:
reverted to ^=

melloware commented:
I got the original implementation from Apache Hive assuming it was correct.  See: https://github.com/apache/hive/blob/master/storage-api/src/java/org/apache/hive/common/util/Murmur3.java

aherbert commented on src/test/java/org/apache/commons/codec/digest/MurmurHash3Test.java:
```diff
@@ -224,4 +224,68 @@ public void testSingleLongRandom() {
+            }
+          }
+        }
+
```
The formatting of these blocks is off. The original code uses tabs. Can you stick to the tab style or use 4 character space indents instead.

MurmurHash2/3 and their tests are the only files in the repo that use tabs. All others use 4 character indentation. This can be fixed at another time.

Claudenw replied:
attempted to fix

garydgregory replied:
In general, we are a no tabs project, see "Spaces No Tabs" here https://commons.apache.org/patches.html. 

It was my mistake to bring in the original PR with tabs in the first place, dang. We can fix tabs separately indeed, whatever makes the process best for review...

melloware replied:
That was my fault I should have scrubbed the tabs when I brought this file over.

aherbert commented on src/main/java/org/apache/commons/codec/digest/MurmurHash3.java:
```diff
@@ -407,22 +415,25 @@ public static long hash64(final byte[] data, final int offset, final int length,
-		long h2 = seed;
+	    // The original algorithm does have a 32 bit unsigned seed.
+	    // We have to mask to match the behavior of the unsigned types and prevent sign extension.
+	    long h1 = seed & UINT_MASK;
```
Indentation issues.

melloware commented:
while we are on the tabs or spaces topic I wonder if Apache commons needs to have a standard formatter for all projects. Similar to this: https://uima.apache.org/codeConventions.html

Where devs can just import it and we can enforce it with a Maven plugin.

Claudenw commented:
I have one that was given to me for commons-collections

On Wed, Nov 6, 2019 at 6:21 PM Melloware <notifications@github.com> wrote:

> while we are on the tabs or spaces topic I wonder if Apache commons needs
> to have a standard formatter for all projects. Similar to this:
> https://uima.apache.org/codeConventions.html
>
> Where devs can just import it and we can enforce it with a Maven plugin.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-codec/pull/27?email_source=notifications&email_token=AASTVHUZ57HVONFMXI6A42TQSMDI3A5CNFSM4JIKUON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDHQEPY#issuecomment-550437439>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASTVHU55WROTGPT7L2YK2LQSMDI3ANCNFSM4JIKUONQ>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

melloware commented:
Is there any way you could attach the formatter to this ticket or point me to where I can get it?

Claudenw commented:
The one I got apparently is from Commons RNG but here it is:



On Thu, Nov 7, 2019 at 12:38 AM Melloware <notifications@github.com> wrote:

> Is there any way you could attach the formatter to this ticket or point me
> to where I can get it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-codec/pull/27?email_source=notifications&email_token=AASTVHQ5OL5HZB3PWWRHRULQSNPOXA5CNFSM4JIKUON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDIPSEI#issuecomment-550566161>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASTVHVWTVLUXCNYIZWZC3TQSNPOXANCNFSM4JIKUONQ>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

melloware commented:
Doesn't look like the attachment came through to GitHub

garydgregory commented:
Each component in Apache Commons has it's own formatting rules :-(
We can start by storing canned configs in each component, for example:
https://gitbox.apache.org/repos/asf?p=commons-dbcp.git;a=blob;f=src/conf/eclipse/formatter.xml;h=e181b4b0e8c03b7d7480cab63d04d94caf383f86;hb=HEAD

Gary

On Thu, Nov 7, 2019 at 2:44 PM Melloware <notifications@github.com> wrote:

> Doesn't look like the attachment came through to GitHub
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-codec/pull/27?email_source=notifications&email_token=AAJB6NZFLZQ64YRZ3UQXTQTQSRVYTA5CNFSM4JIKUON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDNSS4Q#issuecomment-551233906>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJB6N2HHIM325LZ5ERJTILQSRVYTANCNFSM4JIKUONQ>
> .
>

melloware commented:
I like the idea the file should be stored with the project.  And I have used a Maven plugin that you can run the formatter as part of the build or a 1 time job to format the whole project.

aherbert commented:
As the author of the Eclipse formatting configuration in question I would state that I provided it for demo purposes to make formatting a bit easier on a PR. It is by no means a fire and forget configuration. I never use it on an entire file for commons code as it makes no allowances for changing preferences, for example on indentation for method parameters on new lines. Eclipse also mangles HTML in javadoc comments, particularly list elements <li>. So it would not help with a big formatting inconsistency in the javadoc that is often seen across most shared codebases.

I use automated eclipse formatting on other projects with copious amounts of `@formatter:off` tags to prevent formatting code that has been laid out as easy to read. This litters a project with specific tags and I would be wary of doing this for a repository with a lot of contributors with different work set-ups for IDEs, etc.

I like automated tools to do mindless jobs. But code is formatted with the express purpose of being easy to read (compilers don't care) so at least leave the option to allow human formatting to overrule the formatter.

I prefer the option of using a very strict set of checkstyle rules. You can then maintain consistency without having to use an auto-formatter which may rearrange your carefully laid out array declarations, etc. A case in point is that checkstyle for this project should be set up to refuse tabs in source code which arrived in MurmurHash3 and its tests. This would be missed in a PR review as Github converts tabs to whitespace when showing the diff view.

I think a start point would be updating checkstyle to use all the latest rules. Just updating the checkstyle version is not enough, you need to pull in the new rules too.

melloware commented:
@aherbert those are excellent points.  I think I was looking for something because @garydgregory and I want to format all the files in Commons Beanutils2 before its first release.  So we at least start out all of the files in a good state without having to hand fix 200 files.

garydgregory commented:
Going forward, I think we should:

- Store IDE formatting configs in Git
- Encourage use of @formatter:off/on

Gary

On Fri, Nov 8, 2019 at 7:04 AM Melloware <notifications@github.com> wrote:

> @aherbert <https://github.com/aherbert> those are excellent points. I
> think I was looking for something because @garydgregory
> <https://github.com/garydgregory> and I want to format all the files in
> Commons Beanutils2 before its first release. So we at least start out all
> of the files in a good state without having to hand fix 200 files.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-codec/pull/27?email_source=notifications&email_token=AAJB6NZOL654WAGJXFW5E7TQSVIUFA5CNFSM4JIKUON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDQ6EDQ#issuecomment-551674382>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJB6N3A3NBSHTUWIZEJBOLQSVIUFANCNFSM4JIKUONQ>
> .
>

Claudenw commented:
In the projects I work on at work we place the formatting configs for
various systems in a directory called "devtools" (historical reasons).
That directory includes a README.md that describes the shortcomings of the
various formatting config (e.g. Eclipse version does not do X, netbeans
version does not do Y, eclipse requires you enable Z directly from
properties, etc).  This works well for us.

Claude

On Fri, Nov 8, 2019 at 12:51 PM Gary Gregory <notifications@github.com>
wrote:

> Going forward, I think we should:
>
> - Store IDE formatting configs in Git
> - Encourage use of @formatter:off/on
>
> Gary
>
> On Fri, Nov 8, 2019 at 7:04 AM Melloware <notifications@github.com> wrote:
>
> > @aherbert <https://github.com/aherbert> those are excellent points. I
> > think I was looking for something because @garydgregory
> > <https://github.com/garydgregory> and I want to format all the files in
> > Commons Beanutils2 before its first release. So we at least start out all
> > of the files in a good state without having to hand fix 200 files.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/apache/commons-codec/pull/27?email_source=notifications&email_token=AAJB6NZOL654WAGJXFW5E7TQSVIUFA5CNFSM4JIKUON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDQ6EDQ#issuecomment-551674382
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAJB6N3A3NBSHTUWIZEJBOLQSVIUFANCNFSM4JIKUONQ
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-codec/pull/27?email_source=notifications&email_token=AASTVHTGW33XFKK2V2EDJ5TQSVODPA5CNFSM4JIKUON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDRODYQ#issuecomment-551739874>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASTVHT4YOZY3UJSDBESRETQSVODPANCNFSM4JIKUONQ>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

Claudenw commented:
The latest attempt leaves the earlier code in place but deprecates it with a note about the sign extension error.  There is no change to the existing code (despite the change detected by git) except for:

1 the deprecated annotation.
2. the deprecated javadoc tag which says to use the newer version.
3. addition to the 128 bit  method javadocs that indicate the existence of the sign extension error.

New methods are named
hash32_x86and hash128_x64

Addition of UBYTE_MASK, UBYTE_LONG_MASK, and UINT_MASK.

Claudenw commented:
Would someone please merge this?  The two errors  are in a file I did not touch and are in the base build.

aherbert commented on src/test/java/org/apache/commons/codec/digest/MurmurHash3Test.java:
```diff
@@ -224,4 +225,238 @@ public void testSingleLongRandom() {
+        assertEquals(hash1, hash2);
+      }
+
+      public void testStringHash() {
```
Is this missing a `@Test`

aherbert replied:
Also should the method be more explicit: `testStringHashUsesUTF8`

aherbert commented on src/test/java/org/apache/commons/codec/digest/MurmurHash3Test.java:
```diff
@@ -224,4 +225,238 @@ public void testSingleLongRandom() {
+          System.out.println(s);
+          // second time for debugging...
+          hash2 = MurmurHash3.hash32_x86(s, pre, s.length()-pre-post, 123456789);
+        }
```
The repeat of the test should be removed.

It is currently not executed due to a missing `@Test` annotation is the method that uses it.

aherbert commented on src/test/java/org/apache/commons/codec/digest/MurmurHash3Test.java:
```diff
@@ -224,4 +225,238 @@ public void testSingleLongRandom() {
+        }
+    }
+    
+    @Test
```
All code from this point down has different whitespace indentation.

Claudenw replied:
I have created a pull request to change all the tabs to spaces.
https://github.com/apache/commons-codec/pull/28

Plz review

aherbert commented:
There is a lot of new code in here.

IIUC the only method that was broken was the hash32 variant when the input had a number of bytes that was not a factor of 4. This was due to a missing mask to perform sign extension when processing the final 3 bytes. So actually for 1/4 cases of random byte lengths and 1/2 cases for strings it works anyway.

I think the 64-bit and 128-bit versions are not broken as masks are used in the final processing of the last bytes to convert byte to integer then cast to long:

```
@Test
public void testShift() {
    for (int i = 0; i < 256; i++) {
        byte bi = (byte) i;
        for (int j = 0; j < 8; j++) {
            int shift = j * 8;
            // Current method
            long shift1 = (long) (bi & 0xff) << shift;
            // Your new version
            long shift2 = (bi & 0xffL) << shift;
            Assert.assertEquals(shift1, shift2);
        }
    }
}
```

Why deprecate the 128-bit version and provide a whole new implementation that you have copied from elsewhere. Do the two methods produce the same results? Can you call both of them with a few thousand arrays of random bytes with different lengths.

If so then this is just bringing in a new implementation with a different name (hash128 -> hash128_x64).

All of the convenience methods for hash32 are not broken. An int and a long already have a factor of 4 for the bytes. The new methods are just name changes and exactly the same code. So you have a lot of duplication and no need for deprecation other than a name change.

I think the only methods to be deprecated are:
```
public static int hash32(final byte[] data, final int offset, final int length, final int seed)
(plus the 4 methods that call it)

IncrementalHash32.end()
(this can be deprecated in favour of `.end32x86()`)
```

You have added a hash for a CharSequence. Although this would be useful it is not in the scope of this PR which is fixing the broken 32-bit hash implementation.

Am I missing something here?

Claudenw commented:
The 128 bit was broken, the 32 bit was not, the 64 bit is not covered in the original murmurhash implementation.

I brought across the implementation from Yonik Seeley and the tests that go with it.

I duplicated the existing 128 bit tests calling the new 128 hash and and renamed them with "_x86" in the middle (e.g. test128_Double became test128_x64_Double) to verify that all the old tests passed the new hash.

The "testCorrectValues_x64" (which I now realize is misnamed) verifies  that the 32_x86 and 128_x64 return the expected results when compared with the results of the original "C" implementation.
If you modify the testCorrectValues_x86 code to call the original hash32 and hash128 version you will see that the 128 version fails because of the sign extension that occurs early in the process.  (This is what I showed in the original bug) The effective change is at lines 1047 and 1048.

I provided new names for the standard hash methods as there are x86 and x64 versions of the hashes and they do **not** produce the same values.  The change of Hash32 to Hash32_x86 makes it clear which hash it is.  The change from Hash128 to Hash128_x64 does the same. 

While the old Hash128 is incorrectly implemented it has been released and is in the wild, so we have to expect that some uses of the hash128 function can not change to hash128_x64.  To this end the hash128 functions are deprecated (but will probably not be removed) and new implementations are expected to use the hash128_x64 version.

The Hash32 methods are deprecated to refer to the new names.  Given that the code returns the same values one of the implementations could be removed and just call the other.

aherbert commented:
In the case of hash32 I would deprecate those that are known to be broken but not the ones that are not (i.e. the convenience methods). Then provide the new hash32x86 method and delegate calls to a single implementation to eliminate code duplication.

For the hash128x64 method this seems to be the easy solution:

```
// Current
public static long[] hash128(final byte[] data, final int offset, final int length, final int seed) {
    return hash128x64(data, offset, length, seed, seed);
}

// New
public static long[] hash128x64(final byte[] data, final int offset, final int length, final int seed) {
    return hash128x64(data, offset, length, seed & 0xffffffffL, seed & 0xffffffffL);
}

// Single implementation
private static long[] hash128x64(final byte[] data, final int offset, final int length,
        final long h1, final long h2) {
    ...
```

Would that work?

Claudenw commented:
I think you have it backwards.  hash32 is not broken.

hash32 -> hash32_x64 to indicate the implementation in the name (deprecate hash32 infavor of the new name). (code can be coalesced around one implementation with hash32() delegating to hash32_x86() calls.

hash128 is broken.  However there are probably users that can not change from the broken implementation.  I think it is the Cassandra project has the same issue and has elected not to change their hash as it breaks all the existing data sets.  So hash128 deprecated with a note indicating the error in the code so that users can decide if they need to retain the broken one or not.

I will make the changes for Hash32() to call hash30_x86 methods.

Claudenw commented:
There must be a difference between hash32 and hash32_x86 because when I change hash32() to 

     public static int hash32(final byte[] data, final int offset, final int length, final int seed) {
         return hash32_x86( data, offset, length, seed );
    }
The incremental hash test no longer passes.  it fails with 

java.lang.AssertionError: Block size 1 expected:<2094498084> but was:<2002902189>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.apache.commons.codec.digest.MurmurHash3Test.testIncremental(MurmurHash3Test.java:176)
.
.
.

aherbert commented:
The hash32 method is the one with the sign extension error in the finalisation of any remaining bytes. It does not mask the byte with 0xff before the left shift.

Claudenw commented:
and the 128 has a sign extension issue at the start where the original does not use a UINT_MASK., so both implementations have the error.

aherbert commented:
No need for a replacement class for IncrementalHash32. Only the `end()` method of the class should be deprecated. You can just add `hash32x86()` to match the static method name for the same computation.

Can you remove all the underscores from the method names please. That is not Java standard.

Claudenw commented:
Your suggestion not replace Incrementalhash32 means that it will not be possible to use incrementalhash32 to generate hashes that match the old Hash32.  I think we need to be very clear about which methods work with which implementations and to my mind naming of the method is the way to do this.  

We also need to preserve backward compatibility until such time as we make a major version change.

aherbert commented:
In your PR it looks like the `start` and `add` methods of `IncrementalHash32` and `IncrementalHash32_x86` are exactly the same. So just deprecate `end` in favour of a new method `hash32x86`. You can still use the class to generate the old flawed hash but now have a new finalisation method to compute the correct hash.

No need to duplicate the entire class just to change one method.

Claudenw commented:
Are there any other comments on this pull request?  Any open issues?

aherbert commented:
All the deprecated `hash32` method can just call `hash32x86`.

I think you should drop the new implementation of `hash128`. As I previously suggested you can create a private version that is fixed and call it with the two `h1` and `h2` values. The old broken method will not mask them to unsigned. This makes it very clear that the two implementations are basically the same, just the initialisation of the hashes from the seed is broken.

I had a quick test of the methods against Google Guava `com.google.common.hash.Hashing.murmur3_32` and `murmur3_128`. They have chosen to return the values as 16 bytes that are little endian ordered. The output from the the new hash32x86 method matches. Unfortunately (for Google) their `murmur3_128` algorithm has the sign extension error when setting the `h1` and `h2` from the seed. So their implementation matches the broken `hash128` method. I did not do a formal check of the algorithm against [MurmurHash3.cpp](https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp) but checked it by eye. I think this link to the original source should be in the class header. This is the reference for the new implementations, even if adapted from Yonik Seeley. 

The original MurmurHash3 code has an oddity in that the input bytes and returned output byte hash are read/written using the native byte order of the platform. This is little-endian for x86 and x86-64 intel architecture but nothing mandates that this is done. All the computations are done on 64-bit integers which are logically big endian. Perhaps we should state somewhere that the 128-bit hash is returned as two Java longs. To produce the same output as MurmurHash3.cpp would require endianness conversion as appropriate:

```
long[] hash128 = MurmurHash3.hash128x64("hello world");
byte[] bytes = ByteBuffer.allocate(16)
                         .order(ByteOrder.LITTLE_ENDIAN)
                         .putLong(hash128[0])
                         .putLong(hash128[1])
                         .array();

```

There is little in the Javadoc to explain why things are deprecated. All of the helper `hash32` methods that act on primitives that are deprecated are not broken. This is just for a name change. The broken method is the one that processes byte[]. For hash128 the only method that is broken is one that processes byte[] and does not use the default seed. Sign extension of the default seed does not happen as it is positive. The javadoc should explain that the method is broken when a negative seed is used.

If you insist on a new implementation for hash32x86 then it may be more maintainable to copy that to the old `hash32` method and unfix the code to have the bug. Otherwise I would leave hash32 as is and copy it to hash32x86 and fix the sign extension bug. Then add explicit comments in them both above the `// tail` section that this is/is not using the UBYTE_MASK to avoid sign extension in the tail processing.

I also still see a `System.out.print` in the test method `doString`. The whitespace formatting from `testCorrectValues_x64` down is all wrong.

`IncrementalHash32x86` has no comments on any public methods or the class. It uses raw constants for the final mix step when it can call `fmix32`. Same goes for `hash32x86`. 

I would replace inlined rotations with `Integer.rotateLeft` for clarity.

The aim is to reduce code duplication and make this class more maintainable given that it is supporting the same algorithm with and without bugs. I see no point in having two implementations with variations in their style and code. The code in the two methods should be the same except for the bug. It should be very clear to a future maintainer that the code in `hash32` and `hash32x86` and `IncrementalHash32x86` is the same, save for the bug in the finalisation of `hash32` and the `end` function of `IncrementalHash32`. It should also be clear that `hash128` and `hash128x64` are the same except for the initialisation of the hash from the seed.

aherbert commented:
I've been considering the new method `hash32x86(CharSequence data, ...`

I do not think that we should enforce the UTF-8 encoding on the char sequence. A char is a 16-bit unsigned integer. A simpler approach would be to encode the chars as unsigned 16-bit integers. This can process blocks of data just as is done for the byte[] but in blocks of size 2 and not 4. Removing all the if statements and just blindly encoding everything would be simpler. 

How this would effect most cases of ASCII text where the top 8 bits will be empty I do not know. If a user would like to hash using a character set then they should convert to bytes and encode that.

IMO the character set is not relevant. String.equals(Object) treats two Strings as equal if they have the same character in each position. The character set does not matter. It is a binary data comparison. I think hashing should be a binary data operation. It does not care what the data represents (i.e. the character set). The way Guava handles adding a `char` to a hasher is exactly this. It just puts in 2 * 8-bit bytes for each char. The AbstractHasher has these methods:

```
  @Override
  public Hasher putUnencodedChars(CharSequence charSequence) {
    for (int i = 0, len = charSequence.length(); i < len; i++) {
      putChar(charSequence.charAt(i));
    }
    return this;
  }

  @Override
  public Hasher putString(CharSequence charSequence, Charset charset) {
    return putBytes(charSequence.toString().getBytes(charset));
  }
```

So you either add chars as unencoded or you specify a character set which uses a byte conversion.

This makes me think we should change `hashXXX(CharSequence, ...)` to be documented as processing uncoded 16-bit chars and not add new `hashXXX(String [, ...])` methods which are vague. The originals will have to stay but deprecated with an appropriate message. The choice of charsets can be left to the user to convert and use the appropriate byte[] methods.

If you prefer we can drop new methods with String and CharSequence and discuss on the mailing list. This PR can just fix the `byte[]` implementations and clarify naming conventions for the methods.

aherbert commented:
With a bit more time to look at the new CharSequence encoding method I found it to be broken. This test demonstrates how it does not handle bad code points correctly:

```
@Test
public void testHash32x86_UTF8_encoding_InvalidCodePoints() {
    // Since RFC 3629 (November 2003), the high and low surrogate halves used by 
    // UTF-16 (U+D800 through U+DFFF) and code points not encodable by UTF-16 
    // (those after U+10FFFF) are not legal Unicode values, and their UTF-8 
    // encoding must be treated as an invalid byte sequence.
    String invalid = String.valueOf('\uD801');
    int seed = 0;

    // Correct answer
    byte[] bytes = invalid.getBytes(StandardCharsets.UTF_8);
    int h1 = MurmurHash3.hash32x86(bytes, 0, bytes.length, seed);

    // This is NOT EQUAL
    int h2 = MurmurHash3.hash32x86((CharSequence) invalid, 0, invalid.length(), seed);
    Assert.assertNotEquals("Did not expected the encoding to be correct", h1, h2);

    // This is what the encoding should do. Represent bad code points as unknown.
    int h3 = MurmurHash3.hash32x86((CharSequence) "?", 0, 1, seed);
    Assert.assertEquals("The invalid code point should be encoded as '?'", h1, h3);
}
```

I think we should leave charset encoding out of this. 

I'm also confused as to why the default seed has been set at that value. This should be changed for the new methods.

IMO we should have the following new methods and class:

```
public static int hash32x86(byte[] data)
public static int hash32x86(byte[] data, int offset, int length)
public static int hash32x86(byte[] data, int seed)
public static int hash32x86(byte[] data, int offset, int length, int seed)

public static int hash128x64(byte[] data)
public static int hash128x64(byte[] data, int offset, int length)
public static int hash128x64(byte[] data, int seed)
public static int hash128x64(byte[] data, int offset, int length, int seed)

IncrementalHash32x86
```

The default seed for the new methods should be zero. This is what is done in Google Guava and the python library `mmh3`.

This removes adding new methods for primitives since these are not broken. They are just strangely configured helper methods. These should be documented correctly with their equivalent code and scheduled for removal e.g.

```
/**
 * Generates 32 bit hash from two longs with default seed value.
 * This is a helper method that will produce the same result as:
 *
 * <pre>
 * int seed = 104729;
 * int hash = hash32x86(ByteBuffer.allocate(16)
 *                                .putLong(l0)
 *                                .putLong(l1)
 *                                .array(), seed);
 * </pre>
 *
 * @param l0 long to hash
 * @param l1 long to hash
 * @return 32 bit hash
 * @deprecated This method to be removed in 2.0
 */
@Deprecated
public static int hash32(final long l0, final long l1) {
```

The hash64 methods should not be documented as `"MSB 8 bytes of Murmur3 128-bit variant"`. They do not return the same value as and the 128-bit method since they neglect the second hash to be mixed into the first. This has would have an effect as it should be initialised with the default seed (which is not zero). The methods are just plain wrong and they also do not use little endian format for the equivalent calls via a ByteBuffer to the byte[] hash method.

The methods bloat the class. If removed in 2.0 and someone wants them back then this can be addressed then.

I can do documentation changes for all methods to be deprecated in master if you want and then you can rebase the PR to bring in the fixed hash implementations.

Thanks for your patience.

yonik commented:
> I do not think that we should enforce the UTF-8 encoding on the char sequence. A char is a 16-bit unsigned integer. A simpler approach would be to encode the chars as unsigned 16-bit integers.

It looks like this is originally mine, so I'll chime in... the purpose of this method is not a generic string hash, but an optimization for when one wants murmur3 over UTF-8 specifically.  This level of specification is needed for systems that need well defined hashes to enable matching in different languages (typically for things like routing in distributed systems.)

aherbert commented:
Hi @yonik, The code for the UTF-8 hash is nice. But I do not think it should be part of this PR. This PR is specifically trying to fix the current broken implementation of MurmurHash3 for `byte[]` due to some sign extension bugs. It should be simple but it turning into an epic.

I found a bug in your code where it does not support the correct interpretation of illegal characters which should be represented as an invalid byte sequence.

See the example above trying to encode `String.valueOf('\uD801');`

I think the UTF-8 encoding should be part of another PR and ideally fix the support for invalid characters.

aherbert commented:
@Claudenw I have created a separate ticket for the sign extension bug in hash32 [CODEC-267].

I fully documented what the class is currently doing. I rewrote the test class entirely to ensure that all the documentation for what the various methods are doing is tested. This tests hashing with random bytes from the full range of [0-255]. The reference data is from Python mmh3 which is based on the MurmurHash3 c++ code.

I created tests to show the sign extension error.

I then took your fixes for the hash32 bug and hash128 sign extension bug and implemented them in master. The changes log has been attributed to you.

Please can you verify that master passes any additional tests you have written for the PR for the methods with bugs and the new methods, hopefully, without bugs.

The new API has a minimal footprint with 4 new methods and a new IncrementHash32x86 class. I don't think all the other methods are warranted. See [CODEC-268] for a wish to deprecate most of them.

Claudenw commented:
Alex,

Thank you for all your assistance and in getting a final fix in place.  It
passes all the tests I had.  Thank you.

Claude

On Thu, Nov 21, 2019 at 4:13 PM Alex Herbert <notifications@github.com>
wrote:

> @Claudenw <https://github.com/Claudenw> I have created a separate ticket
> for the sign extension bug in hash32 [CODEC-267].
>
> I fully documented what the class is currently doing. I rewrote the test
> class entirely to ensure that all the documentation for what the various
> methods are doing is tested. This tests hashing with random bytes from the
> full range of [0-255]. The reference data is from Python mmh3 which is
> based on the MurmurHash3 c++ code.
>
> I created tests to show the sign extension error.
>
> I then took your fixes for the hash32 bug and hash128 sign extension bug
> and implemented them in master. The changes log has been attributed to you.
>
> Please can you verify that master passes any additional tests you have
> written for the PR for the methods with bugs and the new methods,
> hopefully, without bugs.
>
> The new API has a minimal footprint with 4 new methods and a new
> IncrementHash32x86 class. I don't think all the other methods are
> warranted. See [CODEC-268] for a wish to deprecate most of them.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-codec/pull/27?email_source=notifications&email_token=AASTVHSEMVVSIAB5SCF3XHLQU2XT5A5CNFSM4JIKUON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEE2YT6A#issuecomment-557156856>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASTVHSUYLX5SKT2HLZDKUDQU2XT5ANCNFSM4JIKUONQ>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

aherbert commented:
Thanks for checking. Shall we close this PR?

Claudenw commented:
Does commons-codec publish snapshots?  I don't find them on
https://repository.apache.org/content/repositories/snapshots/ nor any
mention of them on the commons codec page
https://commons.apache.org/proper/commons-codec/

Claude



On Sat, Nov 23, 2019 at 8:38 AM Claude Warren <claude@xenei.com> wrote:

> Alex,
>
> Thank you for all your assistance and in getting a final fix in place.  It
> passes all the tests I had.  Thank you.
>
> Claude
>
> On Thu, Nov 21, 2019 at 4:13 PM Alex Herbert <notifications@github.com>
> wrote:
>
>> @Claudenw <https://github.com/Claudenw> I have created a separate ticket
>> for the sign extension bug in hash32 [CODEC-267].
>>
>> I fully documented what the class is currently doing. I rewrote the test
>> class entirely to ensure that all the documentation for what the various
>> methods are doing is tested. This tests hashing with random bytes from the
>> full range of [0-255]. The reference data is from Python mmh3 which is
>> based on the MurmurHash3 c++ code.
>>
>> I created tests to show the sign extension error.
>>
>> I then took your fixes for the hash32 bug and hash128 sign extension bug
>> and implemented them in master. The changes log has been attributed to you.
>>
>> Please can you verify that master passes any additional tests you have
>> written for the PR for the methods with bugs and the new methods,
>> hopefully, without bugs.
>>
>> The new API has a minimal footprint with 4 new methods and a new
>> IncrementHash32x86 class. I don't think all the other methods are
>> warranted. See [CODEC-268] for a wish to deprecate most of them.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/apache/commons-codec/pull/27?email_source=notifications&email_token=AASTVHSEMVVSIAB5SCF3XHLQU2XT5A5CNFSM4JIKUON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEE2YT6A#issuecomment-557156856>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AASTVHSUYLX5SKT2HLZDKUDQU2XT5ANCNFSM4JIKUONQ>
>> .
>>
>
>
> --
> I like: Like Like - The likeliest place on the web
> <http://like-like.xenei.com>
> LinkedIn: http://www.linkedin.com/in/claudewarren
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

aherbert commented:
Currently codec is not building snapshots. I will raise this on the dev mailing list.

Claudenw commented:
Nit: I think that in MurmurHash3Test line 740 it should use answers2 not
answers:

for (int i = 0; i < answers*2*.length; i++) {

not

for (int i = 0; i < answers.length; i++) {





On Thu, Nov 21, 2019 at 4:13 PM Alex Herbert <notifications@github.com>
wrote:

> @Claudenw <https://github.com/Claudenw> I have created a separate ticket
> for the sign extension bug in hash32 [CODEC-267].
>
> I fully documented what the class is currently doing. I rewrote the test
> class entirely to ensure that all the documentation for what the various
> methods are doing is tested. This tests hashing with random bytes from the
> full range of [0-255]. The reference data is from Python mmh3 which is
> based on the MurmurHash3 c++ code.
>
> I created tests to show the sign extension error.
>
> I then took your fixes for the hash32 bug and hash128 sign extension bug
> and implemented them in master. The changes log has been attributed to you.
>
> Please can you verify that master passes any additional tests you have
> written for the PR for the methods with bugs and the new methods,
> hopefully, without bugs.
>
> The new API has a minimal footprint with 4 new methods and a new
> IncrementHash32x86 class. I don't think all the other methods are
> warranted. See [CODEC-268] for a wish to deprecate most of them.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-codec/pull/27?email_source=notifications&email_token=AASTVHSEMVVSIAB5SCF3XHLQU2XT5A5CNFSM4JIKUON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEE2YT6A#issuecomment-557156856>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASTVHSUYLX5SKT2HLZDKUDQU2XT5ANCNFSM4JIKUONQ>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

aherbert commented:
Yes. I have separated them into tests for positive and negative seed.

Claudenw commented:
Pull request objective fulfilled by other changes to master branch
